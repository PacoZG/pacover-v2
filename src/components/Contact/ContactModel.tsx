'use client'

import emailjs from '@emailjs/browser'
import React, { useState } from 'react'
import { getTheme } from '@/utils/localdb'
import { useField } from '@/hooks/InputHooks'

type EmailValidation = () => boolean
type HandleSendEmail = (event: React.FormEvent<HTMLFormElement>) => void
type DisableButton = () => boolean

export const ContactModel = () => {
  const [disabled, setDisabled] = useState<boolean>(true)
  const [showModal, setShowModal] = useState<boolean>(false)
  const [sending, setSending] = useState<boolean>(false)

  const fullName = useField('text')
  const email = useField('email')
  const message = useField('text')
  const siteKey = process.env.NEXT_PUBLIC_RECAPTCHA_SITE_KEY || ''
  const theme = getTheme()

  const buttonIsDisabled: DisableButton = () => {
    if (!disabled && message.params.value.length > 49) {
      return false
    }
    return true
  }

  const handleVerifyCaptcha = () => {
    setDisabled(!disabled)
  }

  const handleSendEmail: HandleSendEmail = (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    event.preventDefault()
    setSending(true)
    if (
      emailInputValidation() &&
      fullName.params.value.length >= 4 &&
      message.params.value.length >= 50
    ) {
      emailjs
        .sendForm(
          'service_6ked5rp',
          'template_rruy7ci',
          event.target as HTMLFormElement,
          'ayNQSJVuw3mKwJnpn'
        )
        .then(
          result => {
            console.log(result.text)
            setShowModal(true)
            setSending(false)
          },
          error => {
            console.log(error.text)
          }
        )
      fullName.reset()
      email.reset()
      message.reset()
      setTimeout(() => {
        setShowModal(false)
      }, 7000)
    }
  }

  const emailInputValidation: EmailValidation = () => {
    const emailPattern =
      /^(?:[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?\.)+[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-zA-Z0-9-]*[a-zA-Z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])$/
    if (
      emailPattern.test(email.params.value) &&
      email.params.value.length > 11
    ) {
      return true
    }
    return false
  }

  return {
    handleSendEmail,
    emailInputValidation,
    handleVerifyCaptcha,
    buttonIsDisabled,
    disabled,
    showModal,
    sending,
    fullName,
    email,
    message,
    siteKey,
    theme,
  }
}
