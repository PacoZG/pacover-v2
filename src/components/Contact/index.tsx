import React from 'react'
import ReCAPTCHA from 'react-google-recaptcha'
import { Spinner } from '@/components/icons'
import { ContactModel } from '@/components/Contact/ContactModel'
import { useTranslations } from 'next-intl' // Assuming ContactModel is in the same directory

const Contact = () => {
  const t = useTranslations('contact')
  const {
    handleSendEmail,
    handleVerifyCaptcha,
    emailInputValidation,
    buttonIsDisabled,
    showModal,
    sending,
    fullName,
    email,
    message,
    siteKey,
    theme,
  } = ContactModel()

  return (
    <div className="transform duration-500 pt-2 p-4 sm:p-4 pb-0 w-full sm:w-3/5 lg:w-8/12">
      <form onSubmit={handleSendEmail}>
        <div className="flex flex-col items-center justify-center gap-6 p-3 sm:p-5 md:p-7 bg-gray-300 dark:bg-gray-800 rounded-xl shadow-lg transition duration-500">
          {/* Name and Email Inputs */}
          <div className="relative flex flex-col sm:flex-row justify-between items-center gap-6 w-full">
            <input
              className="contact-input-style" // Replaced repetitive class with a custom class
              placeholder={t('name')}
              {...fullName.params}
              required
              name="name"
            />

            <div className="relative flex flex-col w-full">
              <input
                className="contact-input-style" // Replaced repetitive class with a custom class
                placeholder="Email"
                {...email.params}
                required
                name="email"
              />

              {emailInputValidation() ? null : (
                <label className="absolute -bottom-6 left-2 text-sm text-red-700 dark:text-red-400 font-medium">
                  {email.params.value.length > 0 && t('invalid-email')}
                </label>
              )}
            </div>
          </div>

          <textarea
            className="contact-input-style h-24 resize-y"
            placeholder={t('message')}
            {...message.params}
            required
            name="message"
            maxLength={500}
            minLength={50}
          />

          <div className="flex flex-col ml:flex-row ml:items-center justify-evenly ml:justify-between w-full">
            <div className="flex justify-end pt-1 w-full ml:w-auto ml:justify-start">
              {showModal ? (
                <p className="text-sm ml:text-md lg:text-base text-center text-green-700 dark:text-green-400 transform duration-500 font-semibold">
                  {'Thanks / Kiitos / Gracias'}
                </p>
              ) : (
                <label className="text-gray-600 dark:text-gray-400 text-sm ml:text-md lg:text-base text-center sm:text-left">
                  {message.params.value.length > 49 ? (
                    <span className="text-green-700 dark:text-green-400 font-semibold">{`${t(
                      'characters'
                    )}${message.params.value.length}/500 max`}</span>
                  ) : (
                    message.params.value.length > 0 && (
                      <span className="text-red-700 dark:text-red-400 font-medium">{`${t(
                        'characters'
                      )} ${message.params.value.length}/50 min`}</span>
                    )
                  )}
                </label>
              )}
            </div>

            <div className="flex flex-col xl:flex-row items-center justify-center ml:justify-end mt-4 md:mt-0">
              <div className="flex justify-center w-48 xs:w-full m-0">
                <ReCAPTCHA
                  className="scale-75 sm:scale-75 md:scale-80"
                  sitekey={siteKey}
                  onChange={handleVerifyCaptcha}
                  theme={theme}
                />
              </div>

              <div className="flex flex-col items-center ">
                {!showModal ? (
                  <button
                    disabled={buttonIsDisabled()}
                    className={
                      !buttonIsDisabled() || sending === true
                        ? 'contact-send-button-enabled'
                        : 'contact-send-button-disabled'
                    }
                    type="submit"
                  >
                    {!sending ? (
                      t('send')
                    ) : (
                      <div className="flex flex-row items-center gap-3">
                        <Spinner className="animate-spin h-5 w-5 text-current" />

                        <label>{t('sending')}</label>
                      </div>
                    )}
                  </button>
                ) : (
                  <div className="contact-sent-message-display">
                    {t('sent')}
                  </div>
                )}
              </div>
            </div>
          </div>
        </div>
      </form>

      {showModal && (
        <div className="fixed inset-0 bg-black bg-opacity-70 flex items-center justify-center z-50 p-4">
          <div className="bg-white dark:bg-gray-700 rounded-lg p-8 shadow-2xl text-center max-w-sm w-full">
            <p className="text-2xl font-bold text-green-700 dark:text-green-400 mb-4">
              {t('sent')}!
            </p>

            <p className="text-lg text-gray-800 dark:text-gray-200">
              {t('message-sent')}
            </p>
          </div>
        </div>
      )}
    </div>
  )
}

export default Contact
