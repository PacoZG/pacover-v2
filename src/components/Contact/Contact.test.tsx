/* eslint-disabled */
import React from 'react'
import { render } from '@testing-library/react'
import { describe, test, expect, vi, beforeEach } from 'vitest'
import Contact from '@/components/Contact/index'

describe('Contact', () => {
  describe('given configured with dependencies', () => {
    beforeEach(() => {
      window.scrollTo = vi.fn()

      vi.mock('next-intl', () => ({
        useTranslations: () => (key: string) => key,
      }))
    })

    describe('Contact component', () => {
      test('matches the snapshot', () => {
        const { container } = render(<Contact />)

        expect(container).toMatchSnapshot()
      })
    })
  })
})
