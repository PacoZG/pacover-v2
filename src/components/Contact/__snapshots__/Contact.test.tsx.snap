// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Contact > given configured with dependencies > Contact component > matches the snapshot 1`] = `
<div>
  <div
    class="transform duration-500 pt-2 p-4 sm:p-4 pb-0 w-full sm:w-3/5 lg:w-8/12"
  >
    <form>
      <div
        class="flex flex-col items-center justify-center gap-6 p-3 sm:p-5 md:p-7 bg-gray-300 dark:bg-gray-800 rounded-xl shadow-lg transition duration-500"
      >
        <div
          class="relative flex flex-col sm:flex-row justify-between items-center gap-6 w-full"
        >
          <input
            class="contact-input-style"
            name="name"
            placeholder="name"
            required=""
            type="text"
            value=""
          />
          <div
            class="relative flex flex-col w-full"
          >
            <input
              class="contact-input-style"
              name="email"
              placeholder="Email"
              required=""
              type="email"
              value=""
            />
            <label
              class="absolute -bottom-6 left-2 text-sm text-red-700 dark:text-red-400 font-medium"
            />
          </div>
        </div>
        <textarea
          class="contact-input-style h-24 resize-y"
          maxlength="500"
          minlength="50"
          name="message"
          placeholder="message"
          required=""
          type="text"
        />
        <div
          class="flex flex-col ml:flex-row ml:items-center justify-evenly ml:justify-between w-full"
        >
          <div
            class="flex justify-end pt-1 w-full ml:w-auto ml:justify-start"
          >
            <label
              class="text-gray-600 dark:text-gray-400 text-sm ml:text-md lg:text-base text-center sm:text-left"
            />
          </div>
          <div
            class="flex flex-col xl:flex-row items-center justify-center ml:justify-end mt-4 md:mt-0"
          >
            <div
              class="flex justify-center w-48 xs:w-full m-0"
            >
              <div
                class="scale-75 sm:scale-75 md:scale-80"
              />
            </div>
            <div
              class="flex flex-col items-center "
            >
              <button
                class="contact-send-button-disabled"
                disabled=""
                type="submit"
              >
                send
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
